JSON FastStringifier should adapt its buffer capacity based on remaining stack capacity.

MenloDorian opened the pull request:
#### 0762185035ffd66ed5e546b9dad4e97a22dac591
<pre>
JSON FastStringifier should adapt its buffer capacity based on remaining stack capacity.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244323">https://bugs.webkit.org/show_bug.cgi?id=244323</a>
&lt;rdar://problem/99064587&gt;

Reviewed by Darin Adler and Yusuke Suzuki.

FastStringifier relies on remainingCapacity() in m_buffer to limit recursion.

Since it is possible for a JS program to recurse very deep until it is near the bottom of the stack
before calling JSON.stringify(), it is incorrect for FastStringifier to assume that there will always
be enough remaining stack for it to parse and generate sizeof(m_buffer) worth of string.  Instead,
FastStringifier should reduce the amount of useable buffer space based on the amount of remaining
stack capacity available.

In this patch, we do the following:

1. Introduce a m_useableEnd to indicates where in m_buffer we can fill characters up to.
   This is how we reduce the amount of useable buffer space that FastStringifier can use.

2. Adds the computation of m_useableEnd in FastStringifier&apos;s constructor based on remaining
   available stack space.

3. Replace remainingCapacity() with hasRemainingCapacity(), which will:
   a. Check m_length + requested size against m_useableEnd.  Returns true if there&apos;s sufficient capacity.
   b. Else, call hasRemainingCapacitySlow() to recompute m_useableEnd based on actual unused buffer
      space (if necessary), and recompute the capacity check result in (a).
   Without this dynamic adaption of m_useableEnd based on real remaining stack space, Speedometer 2.1
   shows a regression on iOS where stacks are more shallow where the estimate causes us to bail out of
   FastStringifier too eagerly.

Note: we&apos;ve also previously tried a different algorithm where we track a m_cursor and m_cursorStart to
manage where we start filling in content in m_buffer (instead of tracking a m_useableEnd).  That
algorithm requires hasRemainingCapacitySlow() to memmove the filled in portion of m_buffer to shift the
content down every time we need to allow for more capacity in m_buffer.  While this algorithm appears to
be performance neutral, we learn that hasRemainingCapacitySlow() is called many times in the course of
normal operations for any reasonably sized stack.  Hence, we&apos;re potentially doing a lot more memmoves
with this algorithm.

Since both algorithms are performance neutral, we&apos;re going with the one that doesn&apos;t need all the memmoves.

* JSTests/stress/json-fast-stringifier-on-cyclic-structure.js: Added.
(catch):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier::length const):
(JSC::FastStringifier::useableBufferSize):
(JSC::FastStringifier::FastStringifier):
(JSC::FastStringifier::haveFailure const):
(JSC::FastStringifier::result const):
(JSC::FastStringifier::hasRemainingCapacity):
(JSC::FastStringifier::hasRemainingCapacitySlow):
(JSC::FastStringifier::append):
(JSC::FastStringifier::remainingCapacity const): Deleted.

Canonical link: <a href="https://commits.webkit.org/256154@main">https://commits.webkit.org/256154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d4fc61bde273175759ac0b12e727fa0d1135ea4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4074 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104532 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4159 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32255 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100595 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/86067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38667 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81291 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36499 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/28002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40424 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83963 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2030 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42400 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/38819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/18979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me. Please check A/B test to ensure Speedometer2.1 performance is neutral.

darinadler reviewed (commented):
Assuming that we normally do have enough stack space available, I suggest we explore adding a quick check for available stack and skipping the fast stringifier **entirely** if there is insufficient stack space for the worst case, rather than changing the fast stringifying code to run under those conditions.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -662,15 +662,19 @@ class FastStringifier {
     static void logOutcome(ASCIILiteral);
     static void logOutcome(String&&);
 
+    static unsigned useableBufferSize(size_t availableBufferSize);
```
This has a spelling error or typo in the word _usable_. Repeated throughout.

Given that the argument is the number of bytes available in something that has a maximum of 6000, it seems to me that passing a `size_t` is overkill, it would even fit in `uint6_t`.

MenloDorian replied:
I changed "useable" to "usable", and this `size_t` to `unsigned`.  I'm not sure there's much to be gained from using a smaller size.  The codegen on ARM64 might even be slightly less efficient if we go smaller.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -662,15 +662,19 @@ class FastStringifier {
     VM& m_vm;
-    unsigned m_length { 0 };
+    unsigned m_length { 0 }; // length of content already filled into m_buffer.
+    unsigned m_useableEnd { 0 };
```
Ditto.

Typically we call this _capacity_. Maybe that’s a better name for here.

MenloDorian replied:
Yes, `m_capacity` is a better name.  Changed.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -713,10 +717,84 @@ void FastStringifier::logOutcome(String&& outcome)
 
+inline unsigned FastStringifier::useableBufferSize(size_t availableBufferSize)
+{
+    // FastStringifier relies on remaining useable capacity in m_buffer to limit recursion
```
Are you sure this super-long comment is the best way to explain things? Can we tell the story in fewer words?

MenloDorian replied:
There are many details and assumptions (not obvious from the code here) that provide context for the reasoning behind the estimation algorithm.  I think there is value in documenting clearly what these are in case we want to tweak them in the future.  Incidentally, there was also a bug in the comment, which I've now fixed.  I've also reduced a bit of verbosity here and there.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
I would use `auto&` here.

MenloDorian replied:
Changed.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -751,10 +829,28 @@ inline void FastStringifier::recordBufferFull()
+        return false;
+
+    m_useableEnd = m_length + useableSize;
+    ASSERT(m_useableEnd - m_length > size);
```
Should this be `>=`?

MenloDorian replied:
Yes, it should be `>=`.  Fixed.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -662,15 +662,19 @@ class FastStringifier {
+    unsigned m_length { 0 }; // length of content already filled into m_buffer.
+    unsigned m_useableEnd { 0 };
     bool m_checkedObjectPrototype { false };
     bool m_checkedArrayPrototype { false };
```
The comment just below here that says that the `6000` acts a recursion limit is no longer accurate.

MenloDorian replied:
I'll remove the comment below about m_buffer size determining the recursion limit.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -751,10 +829,28 @@ inline void FastStringifier::recordBufferFull()
+{
+    ASSERT(!haveFailure());
+
+    size_t unusedBufferSize = sizeof(m_buffer) - m_length;
```
This doesn’t need to be a `size_t` computation.

MenloDorian replied:
Changed to unsigned.

darinadler reviewed (approved):
Fine as is but I have some refinements you may want to consider.

MenloDorian reviewed (commented):
Thanks for the review.

Constellation reviewed (approved):
r=me too

Commit:
JSON FastStringifier should adapt its buffer capacity based on remaining stack capacity.
https://bugs.webkit.org/show_bug.cgi?id=244323
<rdar://problem/99064587>

Reviewed by Darin Adler and Yusuke Suzuki.

FastStringifier relies on remainingCapacity() in m_buffer to limit recursion.

Since it is possible for a JS program to recurse very deep until it is near the bottom of the stack
before calling JSON.stringify(), it is incorrect for FastStringifier to assume that there will always
be enough remaining stack for it to parse and generate sizeof(m_buffer) worth of string.  Instead,
FastStringifier should reduce the amount of useable buffer space based on the amount of remaining
stack capacity available.

In this patch, we do the following:

1. Introduce a m_useableEnd to indicates where in m_buffer we can fill characters up to.
   This is how we reduce the amount of useable buffer space that FastStringifier can use.

2. Adds the computation of m_useableEnd in FastStringifier's constructor based on remaining
   available stack space.

3. Replace remainingCapacity() with hasRemainingCapacity(), which will:
   a. Check m_length + requested size against m_useableEnd.  Returns true if there's sufficient capacity.
   b. Else, call hasRemainingCapacitySlow() to recompute m_useableEnd based on actual unused buffer
      space (if necessary), and recompute the capacity check result in (a).
   Without this dynamic adaption of m_useableEnd based on real remaining stack space, Speedometer 2.1
   shows a regression on iOS where stacks are more shallow where the estimate causes us to bail out of
   FastStringifier too eagerly.

Note: we've also previously tried a different algorithm where we track a m_cursor and m_cursorStart to
manage where we start filling in content in m_buffer (instead of tracking a m_useableEnd).  That
algorithm requires hasRemainingCapacitySlow() to memmove the filled in portion of m_buffer to shift the
content down every time we need to allow for more capacity in m_buffer.  While this algorithm appears to
be performance neutral, we learn that hasRemainingCapacitySlow() is called many times in the course of
normal operations for any reasonably sized stack.  Hence, we're potentially doing a lot more memmoves
with this algorithm.

Since both algorithms are performance neutral, we're going with the one that doesn't need all the memmoves.

* JSTests/stress/json-fast-stringifier-on-cyclic-structure.js: Added.
(catch):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier::length const):
(JSC::FastStringifier::useableBufferSize):
(JSC::FastStringifier::FastStringifier):
(JSC::FastStringifier::haveFailure const):
(JSC::FastStringifier::result const):
(JSC::FastStringifier::hasRemainingCapacity):
(JSC::FastStringifier::hasRemainingCapacitySlow):
(JSC::FastStringifier::append):
(JSC::FastStringifier::remainingCapacity const): Deleted.

Canonical link: https://commits.webkit.org/256154@main
